ai2thor/controller.py

```diff
@@ -1293,3 +1293,627 @@ def queue_step(self):
+                        assert viz_event.metadata['lastActionSuccess']
+
+                        # Uncomment if you want to view the frames side by side to 
+                        # ensure that they are duplicated.
+                        # from matplotlib import pyplot as plt
+                        # fig = plt.figure()
+                        # fig.add_subplot(2,1,1)
+                        # plt.imshow(self.get_image())
+                        # fig.add_subplot(2,1,2)
+                        # plt.imshow(viz_event.frame)
+                        # plt.show()
+
+                    self.last_event = event
+                    return event
+                     
+        self.last_action_success = False
+        self.last_event.metadata['lastActionSuccess'] = False
+        return self.last_event
+
+
+    def shortest_path(self, source_state, target_state):
+        return self.nx.shortest_path(self.graph, str(source_state), str(target_state))
+    
+    def optimal_plan(self, source_state, target_state):
+        """ This is for debugging. It modifies the state. """
+        path = self.shortest_path(source_state, target_state)
+
+        self.state = source_state
+        actions = []
+        i = 1
+        while i < len(path):
+            for a in self.actions:
+                next_state = self.controller.get_next_state(self.state, a, True)
+                if str(next_state) == path[i]:
+                    actions.append(a)
+                    i += 1
+                    self.state = next_state
+                    break
+        
+        return actions
+
+    def shortest_path_to_target(self, source_state, objId, get_plan=False):
+        """ Many ways to reach objId, which one is best? """
+        states_where_visible = []
+        if self.using_raw_metadata:
+            for s in self.metadata:
+                objects = self.metadata[s]['objects']
+                visible_objects = [o['objectId'] for o in objects if o['visible']]
+                if objId in visible_objects:
+                    states_where_visible.append(s)
+        else:
+            states_where_visible = self.metadata[objId]
+        
+        # transform from strings into states
+        states_where_visible = [
+            self.get_state(*[float(x) for x in str_.split('|')]) for str_ in states_where_visible
+        ]
+
+        best_path = None
+        best_path_len = 0
+
+        for t in states_where_visible:
+            path = self.shortest_path(source_state, t)
+            if len(path) < best_path_len or best_path is None:
+                best_path = path
+                best_path_len = len(path)
+
+        best_plan = []
+        if get_plan:
+            best_plan = self.optimal_plan(source_state, t)
+       
+        return best_path, best_path_len, best_plan
+
+    def visualize_plan(self, source, plan):
+        """ Visualize the best path from source to plan. """
```

Comment:
This comment should be "Visualizes a plan (series of actions) which begins at source."